llvm/test/Feature/Repo/repo_pruning_dependents.py

```diff
@@ -0,0 +1,306 @@
+# Step 6: Generate the Repo IR code from the same ELF IR code.
+#         In this step, the IR code firstly pass the RepoMetadataGenerator pass to generate the
+#         digests of function `a` and variables `str` and `src`. During the RepoPruning pass,
+#         the function `a` is pruned and its dependent `str` is added to the `repo.tickets`.
+# RUN: env REPOFILE=%t.db opt -O0 -mtriple=x86_64-pc-linux-gnu-repo -S %t_5.elf.ll -o %t_5.repo.ll
+# Step 7: Check that 1) the function `a` and it dependent `str` are pruned. 2) the `str` is renamed.
+# RUN: FileCheck %s < %t_5.repo.ll --check-prefix=CHECK-PRUNING
+# Step 8: remove the variable str's TicketNode metadata from the %t_5.repo.2.ll.
+# RUN: python %s -v str -i %t_5.repo.ll -o %t_5.ll
+# Step 9: Generate the Repo object file again.
+#         The repo.tockets in the %t_5.ll file contains the variable `a` and its dependent `str.{{.*}}`.
+#         In this step, the digest of variable `str` is also generated by the RepoObjectWriter but not be
+#         added into the `repo.tickets` since a `str` TicketNode has been added during the pruning pass.
+# RUN: env REPOFILE=%t.db llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj %t_5.ll -o %t_5.o
+# Step 10: Dump the compilation members and check the "name:str"
+# RUN: env REPOFILE=%t.db pstore-dump --all-compilations %t.db | FileCheck %s
+
+def write_elf_IR_test4 (filename):
+  """
+  Writes an IR test4 file.
+  """
+  with open(filename, 'w') as f:
+    f.write ('target triple = "x86_64-pc-linux-gnu-elf"\n\n')
+    f.write ('@str = private constant [6 x i8] c"Ipsum\\00", align 1\n\n')
+    f.write ('define void @g() {\n')
+    f.write ('entry:\n')
+    f.write ('  %puts = tail call i32 @puts(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @str, i64 0, i64 0))\n')
+    f.write ('  ret void\n')
+    f.write ('}\n\n')
+    f.write ('declare i32 @puts(i8* nocapture readonly)\n')
+
+
+# CHECK-PRUNING: !TicketNode(name: "a", digest: [16 x i8] c"{{.*}}", linkage: external, visibility: default, pruned: true)
+# CHECK-PRUNING:!TicketNode(name: "str.{{.*}}", digest: [16 x i8] c"{{.*}}", linkage: private, visibility: default, pruned: true)
+
+def write_elf_IR_test5 (filename):
+  """
+  Writes an IR test5 file.
+  """
+  with open(filename, 'w') as f:
+    f.write ('target triple = "x86_64-pc-linux-gnu-elf"\n\n')
+    f.write ('@str = private constant [6 x i8] c"Ipsum\\00", align 1\n')
+    f.write ('@src = global i8* getelementptr inbounds ([6 x i8], [6 x i8]* @str, i32 0, i32 0), align 4\n\n')
+    f.write ('define void @a() {\n')
+    f.write ('entry:\n')
+    f.write ('  %puts = tail call i32 @puts(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @str, i64 0, i64 0))\n')
+    f.write ('  ret void\n')
+    f.write ('}\n\n')
+    f.write ('declare i32 @puts(i8* nocapture readonly)\n')
+
+# CHECK:     name    : str
+# CHECK:     name    : str
+# CHECK-NOT: name    : str
+
+def remove_function_ticketnode (functions, inputfile, outputfile):
```

Comment:
These functions remove the reference to the ticketnode:

`define i32 @A() !repo_ticket !0 {`  => `define i32 @A() {`

But don't remove the TicketNode itself, which still refers to the function by name:

`!repo.tickets = !{!0}	

!0 = !TicketNode(name: "A", digest: [16 x i8] c"\BC\B0\9B/\86\B8\09I\F7P\FB\1D\DD\F1Kf", linkage: external, visibility: default, pruned: false)`

Is having that metadata okay, or could it potentially affect the test? These are harder to remove as, IIRC, LLVM wants metadata to be sequentially numbered.

Reply:
> 
> Is having that metadata okay, or could it potentially affect the test? 
>
It is ok and doesn’t affect the test. Thinking about the following IR code:
```
target triple = "x86_64-pc-linux-gnu-repo"

define i32 @C()  {
entry:
  ret i32 1
}

!repo.tickets = !{!0, !1, !2}

!0 = !TicketNode(name: "A", digest: [16 x i8] c"\8F\DD\E8\BC]*.e\B5F]r\8E\8CA\8F", linkage: external, visibility: default, pruned: true)
!1 = !TicketNode(name: "B", digest: [16 x i8] c"\19}\B9|\08\16Y\8D\E0\C1G\974+S\D9", linkage: external, visibility: default, pruned: true)
!2 = !TicketNode(name: "C", digest: [16 x i8] c"\0E\F8K\12\FA#y\F6\F7\97\C6\A7u\1A\89$", linkage: external, visibility: default, pruned: false)
```

Since the TicketNode `0` and `1` have the `true` pruned flag, they are used to generate the compilation members in the backend. However, the TicketNode `2` is not referenced by the definition of the function `C` and has a `false` pruned flag, it is not used by the backed. Therefore, it doesn’t affect the test

Reply:
Okay, thanks.